boards/PDM/Src/Io/Io_Efuse.c

```diff
@@ -0,0 +1,847 @@
+    uint16_t      ChipSelect_GPIO_Pin,
+    struct Efuse *e_fuse)
+{
+    ExitCode exit_code;
+    uint16_t command   = 0x0000U;
+    uint16_t tx_data[] = { 0x0000U };
+
+    // Place the Status Register address into bits 10->13
+    command =
+        (uint16_t)((SI_STATR_0_ADDR & EFUSE_ADDR_MASK) << EFUSE_ADDR_SHIFT);
+    // Shift bit 3 of the address (SOA3: the channel number) to the 13th bit
+    command = (uint16_t)(
+        command | ((register_address & SOA3_MASK) << EFUSE_ADDR_SHIFT) |
+        (register_address & (SOA2_MASK | SOA1_MASK | SOA0_MASK)));
+    // Invert watchdog bit state (Note: It is safe to do so even if the watchdog
+    // is disabled)
+    if (e_fuse->wdin_bit_to_set)
+    {
+        command |= WATCH_DOG_BIT;
+    }
+    else
+    {
+        command &= ~(WATCH_DOG_BIT);
+    }
+    // Invert watchdog bit state for next write
+    e_fuse->wdin_bit_to_set = !e_fuse->wdin_bit_to_set;
+
+    // Compute and set/clear parity bit
+    Io_Efuse_SetParityBit(&command);
+
+    tx_data[0] = command;
+    exit_code  = Io_Efuse_ReadFromEfuse(
+        tx_data, register_value, ChipSelect_GPIO_Port, ChipSelect_GPIO_Pin);
+    // Only return register contents and clear bits 9->15
+    CLEAR_BIT(*register_value, ~EFUSE_SO_DATA_MASK);
+
+    return exit_code;
+}
+
+static void Io_Efuse_SetParityBit(uint16_t *spi_command)
+{
+    uint16_t spi_data = *spi_command;
+    uint8_t  parity_bit;
+    // Calculate parity for bits in spi command
+    for (parity_bit = 0; spi_data > 0; spi_data >>= 1)
+    {
+        parity_bit ^= (spi_data & 1);
+    }
+
+    if (parity_bit == 1)
+    {
+        SET_BIT(*spi_command, PARITY_BIT);
+    }
+    else
+    {
+        CLEAR_BIT(*spi_command, PARITY_BIT);
+    }
+}
+
+static ExitCode Io_Efuse_WriteToEfuse(
+    uint16_t *    TxData,
+    GPIO_TypeDef *GPIO_Port,
+    uint16_t      GPIO_Pin)
+{
+    HAL_StatusTypeDef status;
+
+    HAL_GPIO_WritePin(GPIO_Port, GPIO_Pin, GPIO_PIN_RESET);
+    status = HAL_SPI_TransmitReceive(
+        efuse_spi_handle, (uint8_t *)TxData, (uint8_t *)TxData, 1U, 100U);
+    HAL_GPIO_WritePin(GPIO_Port, GPIO_Pin, GPIO_PIN_SET);
+
+    if (status != HAL_OK)
+    {
+        return EXIT_CODE_TIMEOUT;
+    }
+
+    return EXIT_CODE_OK;
+}
+
+static ExitCode Io_Efuse_ReadFromEfuse(
+    uint16_t *    tx_data,
+    uint16_t *    rx_data,
+    GPIO_TypeDef *chip_select_port,
+    uint16_t      chip_select_pin)
+{
+    HAL_StatusTypeDef status;
+
+    // Send command to read from status register
+    // Data is returned on the following SPI transfer
+    HAL_GPIO_WritePin(chip_select_port, chip_select_pin, GPIO_PIN_RESET);
+    status = HAL_SPI_TransmitReceive(
+        efuse_spi_handle, (uint8_t *)tx_data, (uint8_t *)rx_data, 1U, 100U);
+    HAL_GPIO_WritePin(chip_select_port, chip_select_pin, GPIO_PIN_SET);
+
+    if (status != HAL_OK)
+    {
+        return EXIT_CODE_TIMEOUT;
+    }
+
+    // Receive data from E-fuse by sending dummy data
+    *tx_data = 0xFFFF;
```

Comment:
Do we still need if we are using HAL_SPI_Receive ?

Reply:
Discussed this with Ken and after checking the how the SPI peripheral works we determined that even if we are only sending data and we don't care about the data sent back from the slave we still need to read it to de-queue the SPI Rx buffer for the next read. `HAL_SPI_Transmit()` should only ever be used if you never plan to use `HAL_SPI_Receive()` or `HAL_SPI_TransmitReceive()` after as the Rx is not flushed till the data is read-out. Ie. calling `HAL_SPI_Receive()` after `HAL_SPI_Transmit()` would give you the data sent by the slave during the `HAL_SPI_Transmit()` call.

![image](https://user-images.githubusercontent.com/25499626/87756016-fbc51e00-c7bc-11ea-894a-fb11ca55a850.png)

https://www.st.com/content/ccc/resource/training/technical/product_training/group0/52/17/7a/28/2b/90/41/7d/STM32H7-Peripheral-Serial_Peripheral_interface_SPI/files/STM32H7-Peripheral-Serial_Peripheral_interface_SPI.pdf/_jcr_content/translations/en.STM32H7-Peripheral-Serial_Peripheral_interface_SPI.pdf

Reply:
Ok thanks this makes sense 👍

Reply:
We sure miss you Yoshi man ...... we sure miss you